### PR TITLE
feat: Continuous scheduling and analytics table job

### PR DESCRIPTION
**Pull request by larshelge**
* Introduces scheduling of jobs with _fixed delay_, i.e. jobs that run continuously with a specified number of seconds delay in between the executions. This includes a new `SchedulingType` enum for `JobType`, which indicates whether the job is triggered by cron expression or with a fixed delay between executions. There is a 10 seconds default delay between when _fixed delay_ jobs are scheduled and invoked the first time. 

* Introduces a continuous analytics table job, which updates the _latest_ analytics table partition continuously, and updates the full analytics table once per day.

* The latest analytics table delay is specified in seconds using the `delay` property on `JobConfiguration`. The time of the full analytics table update is specified using the `hourOfDay` property on `ContinuousAnalyticsJobParameters`.

### Testing

To test this PR, you can create a continuous analytics table job with a request like this:

    POST /api/jobConfigurations

Headers:

    Content-Type: application/json

Payload:

```json
{
    "name": "Continuous analytics table",
    "jobType": "CONTINUOUS_ANALYTICS_TABLE",
    "delay": 5,
    "enabled": true,
    "jobParameters": {
    	"hourOfDay": 1
    }
}
```

### CODE REVIEW

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/JobConfiguration.java [hunk 1]
```diff
@@ -77,10 +78,46 @@
...
+    /**
+     * The delay in seconds between the completion of one job execution and the
+     * start of the next. Relevant for scheduling type {@link SchedulingType#FIXED_DELAY}.
+     */
+    private Integer delay;
+
+    /**
+     * Indicates this job should be triggered continuously.
+     */
+    private boolean continuousExecution = false;
```
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/JobConfiguration.java]: This property is just moved within the class, not added.

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/JobConfigurationObjectBundleHook.java [hunk 2]
```diff
@@ -279,7 +279,7 @@ private JobConfiguration validatePersistedAndPrepareTempJobConfiguration( List<E
     private void validateJobConfigurationWithNonContinuousExecution( List<ErrorReport> errorReports,
         JobConfiguration jobConfiguration )
     {
-        if ( !jobConfiguration.isContinuousExecution() )
+        if ( !jobConfiguration.isContinuousExecution() && jobConfiguration.getJobType().isCronSchedulingType()  )
```
Comment (stian-sandvold) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/JobConfigurationObjectBundleHook.java]:
```suggestion
        if ( !jobConfiguration.isContinuousExecution() && jobConfiguration.getJobType().isCronSchedulingType() )
```

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/JobConfiguration.java [hunk 3]
```diff
@@ -79,13 +83,23 @@
...
+     * The cron expression used for scheduling the job. Relevant for scheduling
+     * type {@link SchedulingType#CRON}.
+     */
+    private String cronExpression;
+
+    /**
+     * The delay in seconds between the completion of one job execution and the
+     * start of the next. Relevant for scheduling type {@link SchedulingType#FIXED_DELAY}.
+     */
+    private Integer delay;
```
Comment (stian-sandvold) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/JobConfiguration.java]: We also have a property in JobConfiguraition: "continuousExecution", which right now just reschedules the job after it finishes. I guess this PR removes the need for that property?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/parameters/ContinuousAnalyticsJobParameters.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+ * @author Lars Helge Overland
+ */
+@JacksonXmlRootElement( localName = "jobParameters", namespace = DxfNamespaces.DXF_2_0 )
+@JsonDeserialize( using = ContinuousAnalyticsJobParametersDeserializer.class )
+public class ContinuousAnalyticsJobParameters
+    implements JobParameters
+{
+    private static final long serialVersionUID = 4613054056442276592L;
+
+    private Integer hourOfDay;
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/parameters/ContinuousAnalyticsJobParameters.java]: I'd add a short javadoc to explain what's the purpose of these properties

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/table/scheduling/ContinuousAnalyticsTableJob.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+            AnalyticsTableUpdateParams params = AnalyticsTableUpdateParams.newBuilder()
+                .withLastYears( parameters.getLastYears() )
+                .withSkipResourceTables( false )
+                .withSkipTableTypes( parameters.getSkipTableTypes() )
+                .withJobId( jobConfiguration )
+                .withStartTime( now )
+                .build();
+
+            analyticsTableGenerator.generateTables( params );
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/table/scheduling/ContinuousAnalyticsTableJob.java]: If an exception is thrown here, the `update` variable will not be updated and, I think, the scheduler will stop working?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/JobConfiguration.java (hunk 3 above)
Comment (larshelge) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/JobConfiguration.java]: Yes correct. I will create a new PR for removal of that property once this is merged.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/parameters/ContinuousAnalyticsJobParameters.java (hunk 4 above)
Comment (larshelge) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/parameters/ContinuousAnalyticsJobParameters.java]: Added Javadoc now.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/table/scheduling/ContinuousAnalyticsTableJob.java (hunk 5 above)
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/table/scheduling/ContinuousAnalyticsTableJob.java]: Good point. Added a try-finally clause.

**Pull request merged**